### ISSUE TITLE
Idea for a rule improvement: Duplicate Map (Number) Literal - exclude block/closure by pattern

### ISSUE THREAD

**Issue by VitaliiH** [CodeNarc/CodeNarc#175]
We are using Groovy and Grails rules together with SonarCube for our Grails application. The mentioned rules work correct. Unfortunately, on Grails domains they are triggered in undesirable way for a specific cases. A Grails domain can have a "constaints" block, to eliminate restrictions for a domain fields:
`static constraints = {` 
`firstName nullable: false, blank: false, size: 15`
`lastName nullable: false, blank: false, size: 15`
`}`
These two lines will cause violations for both rules. This is correct, but doesn't make sence for "constraints" block. If I extract a map into a variabe, whole block will be looking weird. I even can't find suitable name for a variable.
I've looked through Sonar configuration but didn't find anything useful for me, how to avoid undesirable triggering of the rules. Currently I don't take into account marking blocks with "SuppressWarinings" annotation, because my issue may also be useful for other cases and frameworks. Also, currently we have over 300 domains, I don't really want to do that on all of them.

So my question is, wheter I understand configuration of rules correctly and there is no way to exclude some blocks/methods for mentioned rules. I saw similar configs for other rules, but not for these.

On my opinion, it would be great to have an ability to set "exceptional" blocks for these rules.
Another way - create similar rules with mentioned exceptions under Grails rule set.

Would be thankful for any suggestion or discussion.

**Comment by chrismair** [comment]
CodeNarc has no in-depth knowledge of Grails and its artifacts, so it must rely on the class/file name-based facilities for configuring/suppressing rules, e.g. those described at:
http://codenarc.sourceforge.net/codenarc-configuring-rules.html#Standard_Properties_for_Configuring_Rules

Are your domain classes in a particular package or folder, or do they follow a class-naming convention? And I guess the question is also which of those facilities are exposed when running through SonarCube.

**Comment by VitaliiH** [comment]
Unfortunately, this configuration excludes a whole file while I want left rest of file still to be checked.
Also, looks like we can't configure the default attributes in SonarQube, but it's different question.
Another important thing I forgot about - in Grails it's also possible to hav DTOs with the same 
`static constraints` block.
It is used for the same purpose and is exposed to the same problem.

I found that SonarQube's configuration "Ignore Issues in Blocks" may work for me.
In case if anybody curious:
Explanation of the config: http://docs.sonarqube.org/display/SONAR/Narrowing+the+Focus
I added `static constraints` as "start of block" parameter and `\}` as "end of block" parameter. As these blocks contains anything else but maps, I'm sure it's not a problem to exclude them at all.
The same trick I'm going to apply for `static mapping` blocks which are pretty the same.

Chris, Thank you a lot for your answer, it helped me to understand I didn't miss anything simple and important.
